test/subspace-test-service/src/lib.rs

```diff
@@ -528,11 +528,14 @@ impl MockPrimaryNode {
     }
 
     /// Remove a ready transaction from transaction pool.
-    pub fn prune_tx_from_pool(&self, tx: &OpaqueExtrinsic) -> Result<(), Box<dyn Error>> {
+    pub async fn prune_tx_from_pool(&self, tx: &OpaqueExtrinsic) -> Result<(), Box<dyn Error>> {
         self.transaction_pool.pool().prune_known(
             &BlockId::Hash(self.client.info().best_hash),
             &[self.transaction_pool.hash_of(tx)],
         )?;
+        // `ban_time` have set to 0, explicitly wait 1ms here to ensure `clear_stale` will remove
+        // all the bans as the ban time must be passed.
+        tokio::time::sleep(time::Duration::from_millis(1)).await;
```

Comment:
Due to async nature of the matter, there is zero guarantee that 0ms will happen before 1ms delay!

Reply:
Not knowledgeable on this point honestly. Does it help with a larger value like 100ms? I mean, some parameter that works practically.

Reply:
I'd say this should be checking some objective fact that can be proven to not have race condition. Even 100ms is technically not guaranteed to happen before 0ms if the thread that is supposed to run 100ms happens to be stuck with something compute-heavy for 100ms.

The approach is fundamentally flawed IMO, but I have no context for the use case.

Reply:
That is okay, the bans are not asynchronously removed in the background but instead in the call of `clear_stale`, where it will compare the previously recorded `banned_until` with `Instant::now()` to decide whether remove the ban.

Reply:
If it is not asynchronous then why do you need a delay?

Reply:
When a tx is baned, its `banned_until` is recorded as `Instant::now() + cfg.ban_time` (`ban_time` is set to 0 in test), inside `clear_stale`  if `banned_until < Instant::now()` is true then the ban will be removed, add a delay can ensure the ban is expired.

Reply:
You understand that `Instant::now()` is not guaranteed to be any higher than previous `Instant::now()` regardless of how much time have passed, right? It will not be less, but it may easily be exactly the same.

Reply:
Yes, I know it is a  monotonic clock and it only guarantees to be nondecreasing. 

> but it may easily be exactly the same

It won't be that easy I think, it should be the rare case and will be less likely to happen after explicitly waiting for 1ms. 

Although it is still not a guarantee but should be enough for the test since even for now the test is not breaking too often.

Reply:
If you think it is sufficient - sure, I just don't like probabilistic stuff like this in general

Reply:
Let's see how it helps in practice.